PIVOT Bugs Fix

RIC06X opened the pull request:
### Description

Previously, Babelfish PIVOT operator has many bugs and limitation. In this PR, we fixed the bugs and blocked some unsupported edge cases.

### Issues Resolved
BABEL-284-bug-fix (bugs tracker issue)
BABEL-4567, BABEL-4568, BABEL-4569, BABEL-4576, BABEL-4558, BABEL-4559
* fixed pivot used in function/procedure will cause server restart
* fixed table/schema name used in 'FOR (colname)' will cause server restart
* fixed view usage in pivot stmt produce empty result
* NULL should be the aggregate result when no row is selected(except COUNT)
* Error message for creating view on pivot, 
* Error message when use pivot in WITH CTE 
* Error message when use join with PIVOT stmt

### Test Scenarios Covered ###
- call function/procedure multiple times
- table schema name used in 'FOR (colname)'
- view usage in pivot stmt
- NULL value output when no row is selected by aggregate function in PIVOT
- creating view on pivot
- use pivot in WITH CTE
- join with PIVOT stmt
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

RIC06X committed:
[BABEL-284] Support table name prefix in pivot for column

Support table name prefix in pivot for column

Task: BABEL-284
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-284] Fix pivot function/procedure bug

Fixed an issue that pivot used in function/procedure will cause server
restart

Task: BABEL-284
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
[BABEL-4558] Fixed double analyzing caused issue

Previously we double parsed fromClause in pivot stmt and this is causing
the "unrecognized node type" error msg. This commit resolved this bug by
using a fresh copy of fromClause in pivot_src_sql.

This commit also corrected some of the pivot test results.

Task: BABEL-4558
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3839,17 +3840,19 @@ bbf_pivot(PG_FUNCTION_ARGS)
 	}
 	PG_END_TRY();
-
+	MemoryContextSwitchTo(oldcontext);
```
It won't switch back to oldContext if it catch an excption ?

RIC06X replied:
Yes, we could add `MemoryContextSwitchTo(oldcontext);` at line 3852. I'm curious if we throw an error should we also take care of restoring to the oldcontext?

forestkeeper replied:
In this case we should in this case
We're trying to follow Defensive programming.

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3892,7 +3896,7 @@ bbf_pivot(PG_FUNCTION_ARGS)
 
 	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
-	tsql_outmost_estat->pivot_parsetree_list = list_delete_nth_cell(tsql_outmost_estat->pivot_parsetree_list, tsql_outmost_estat->pivot_number - 1);
+	tsql_outmost_estat->pivot_parsetree_list = list_delete_nth_cell(tsql_outmost_estat->pivot_parsetree_list, 0);
```
What does change to 0 in here mean?

RIC06X replied:
The execution order is first one in the list get executed first and that's why we change from last to 0. After execution, we will remove the head element of the list.

forestkeeper commented:
- Please mention the bug JIRA, not the original pivot feature jira in the commit
- Add test cases for each bug fix

RIC06X committed:
Added error messages for unsupported edge cases

This commit added error message for creating view on pivot, use pivot in
WITH CTE, and join with PIVOT stmt.

Task: BABEL-4567, BABEL-4569, BABEL-4576
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
Fix bugs and add test cases

This commit resolved the following issues:
 - view usage in pivot stmt produce empty result
 - NULL should be the aggregate result when no row is selected(except
    COUNT)
This commit also added the following test cases:
 - join usage in pivot stmt
 - view usage in pivot stmt
 - aggregate string values with NULL output result for pivot stmt

Task: BABEL-4558
Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

RIC06X committed:
Add pivot struct to improve code readability

Added tsql_pivot_fields stuct to store pivot related information to
improve readability

Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3776,7 +3778,7 @@ SPI_execute_raw_parsetree(RawStmt *parsetree, bool read_only, long tcount)
 	 * it is not important here
```
I think this comment needs to be changed, since last time you're not using sourcetext in here

RIC06X replied:
Yes, will do

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3991,10 +4001,12 @@ load_categories_hash(RawStmt *cats_sql, MemoryContext per_query_ctx)
+						const char	*funcName,
+						HTAB 		*bbf_pivot_hash,
+						TupleDesc 	tupdesc,
+						bool 		randomAccess)
```
What does randomAccess do in this function ?

RIC06X replied:
randomAccess: if true, both forward and backward accesses to the tuple store are allowed.

SFRM_Materialize mode expects us to return a NULL Datum. The actual tuples are in our tuplestore and passed back through rsinfo->setResult. rsinfo->setDesc is set to the tuple description that we actually used to build our tuples with, so the caller can verify we did what it was expecting.

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -4029,6 +4041,8 @@ get_bbf_pivot_tuplestore(RawStmt *sql,
 		int			non_pivot_columns;
 		int			result_ncols;
+		/* only COUNT will output 0 when no row is selected */
+		bool		output_zero = !strcmp(funcName, "count");
```
Did you consider "COUNT" ?

forestkeeper replied:
Could you paste the PG doc url for this comment too

RIC06X replied:
bbf on the first place(during bison parser) will lower all function name. I'll change to strcasecmp to prevent name in mixed case.

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3930,7 +3937,7 @@ load_categories_hash(RawStmt *cats_sql, MemoryContext per_query_ctx)
 	/* Retrieve the category name rows */
-	ret = SPI_execute_raw_parsetree(cats_sql, true, 0);
+	ret = SPI_execute_raw_parsetree(cats_sql, sourcetext, true, 0);
 	proc = SPI_processed;
```
It'll better to change this var name to tuple_processed.

Proc has some other meaning in the code and may cause misunderstanding.

RIC06X replied:
yes, will change the var name

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3950,26 +3957,29 @@ load_categories_hash(RawStmt *cats_sql, MemoryContext per_query_ctx)
-			catdesc->catname = catname;
+			catdesc->catname = pstrdup(catname_lower);
 			catdesc->attidx = i;
 			/* Add the proc description block to the hashtable */
```
I think it better use tuple description block in the comment

proc in PG context normally mean procedure.

RIC06X replied:
yes, will change the name

RIC06X committed:
Update var name and comments

Updated var name and added new comments

Signed-off-by: Yanjie Xu <yanjiex@amazon.com>

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3839,17 +3843,20 @@ bbf_pivot(PG_FUNCTION_ARGS)
 	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
```
Inside the pg try we didn't copy or do any memory allocate, why we switch to outmost context at here ?

RIC06X replied:
Removed PG_TRY and MemoryContextSwitch code.

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3892,7 +3901,7 @@ bbf_pivot(PG_FUNCTION_ARGS)
 	MemoryContextSwitchTo(oldcontext);
 
 	oldcontext = MemoryContextSwitchTo(tsql_outmost_context);
```
Why not directly switch to tsql_outmost_context in here ? Looks equal.

RIC06X replied:
removed that line. Resolved

forestkeeper commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -3950,28 +3959,31 @@ load_categories_hash(RawStmt *cats_sql, MemoryContext per_query_ctx)
 						 errmsg("provided \"categories\" SQL must " \
 								"not return NULL values")));
 			
 			SPIcontext = MemoryContextSwitchTo(per_query_ctx);
```
Pls use oldMemoryContext for the var name in here, the SPIcontext has other meaning for var naming.

RIC06X replied:
Renamed in latest commit.

RIC06X committed:
Rename variables and delete unnecessary code

Renamed variables and deleted unnecessary code.

Signed-off-by: Yanjie Xu <yanjiex@amazon.com>
